src/bio/bio_recovery.c

```diff
@@ -0,0 +1,250 @@
+/**
+ * (C) Copyright 2018-2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. B620873.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+#define D_LOGFAC	DD_FAC(bio)
+
+#include <spdk/io_channel.h>
+#include <spdk/blob.h>
+#include "bio_internal.h"
+
+/*
+ * The BIO blobstore (mapped to one NVMe device) owner xstream polls the
+ * state periodically, and takes predefined reaction routines on state
+ * transition, all the reaction routines must be non-blocking, otherwise
+ * the DAOS progress ULT will be blocked, and NVMe device qpair won't be
+ * polled.
+ */
+struct bio_reaction_ops	*ract_ops;
+
+/*
+ * Return value:	0: Faulty reaction is done;
+ *			1: Faulty reaction is in progress;
```

Comment:
what would happen if the faulty reaction were to still be in process (return 1)? since it is non-blocking, it would just keep going, then eventually check at a later time if complete? or I guess bio_bs_state_transit() would just be called again since a return code of 1 would not change the state?

Reply:
Right, the bio_bs_state_transit() will be called by monitor periodically.